src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -419,4 +418,128 @@ public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable Index
+        if (canDeleteShardContent(shardId, indexSettings) == false) {
+            throw new ElasticsearchIllegalStateException("Can't delete shard " + shardId);
+        }
+        logger.info("{} deleting shard reason [{}]", shardId, reason);
+        nodeEnv.deleteShardDirectoryUnderLock(lock, indexSettings);
+    }
+
+    /**
+     * This method deletes the shard contents on disk for the given shard ID. This method will fail if the shard deleting
+     * is prevented by {@link #canDeleteShardContent(org.elasticsearch.index.shard.ShardId, org.elasticsearch.cluster.metadata.IndexMetaData)}
+     * of if the shards lock can not be acquired.
+     * @param reason the reason for the shard deletion
+     * @param shardId the shards ID to delete
+     * @param metaData the shards index metadata. This is required to access the indexes settings etc.
+     * @throws IOException if an IOException occurs
+     */
+    public void deleteShardStore(String reason, ShardId shardId, IndexMetaData metaData) throws IOException {
+        final Settings indexSettings = buildIndexSettings(metaData);
+        if (canDeleteShardContent(shardId, indexSettings) == false) {
+            throw new ElasticsearchIllegalStateException("Can't delete shard " + shardId);
+        }
+        logger.info("{} deleting shard reason [{}]", shardId, reason);
+        nodeEnv.deleteShardDirectorySafe(shardId, indexSettings);
+    }
+
+    /**
+     * Returns <code>true</code> iff the shards content for the given shard can be deleted.
+     * If the shard is still allocated / active on this node or for instance if the shard is located on shared
+     * filesystem this method will return <code>false</code>.
+     *
+     * @param shardId the shard to delete.
+     * @param metaData the shards index metadata. This is required to access the indexes settings etc.
+     */
+    public boolean canDeleteShardContent(ShardId shardId, IndexMetaData metaData) {
+        assert shardId.getIndex().equals(metaData.getIndex());
+        final Settings indexSettings = buildIndexSettings(metaData);
+        return canDeleteShardContent(shardId, indexSettings);
+    }
+
+    private boolean canDeleteShardContent(ShardId shardId, @IndexSettings Settings indexSettings) {
+        final Tuple<IndexService, Injector> indexServiceInjectorTuple = this.indices.get(shardId.getIndex());
+        // TODO add some protection here to prevent shard deletion if we are on a shard FS or have ShadowReplicas enabled.
+        if (indexServiceInjectorTuple != null && nodeEnv.hasNodeFile()) {
+            final IndexService indexService = indexServiceInjectorTuple.v1();
+            return indexService.hasShard(shardId.id()) == false;
+        } else if (nodeEnv.hasNodeFile()) {
+            final Path[] shardLocations = nodeEnv.shardPaths(shardId);
+            return FileSystemUtils.exists(shardLocations);
+        }
+        return false;
+    }
+
+    private Settings buildIndexSettings(IndexMetaData metaData) {
```

Comment:
there are node settings and index settings which both we need. since in some nodes you might have a different index dir etc. I think its crucial since we might have per node settings ie. to prevent shard deleteion in the future.